presto-plugin-toolkit/src/main/java/io/prestosql/plugin/base/security/FileBasedAccessControl.java

```diff
@@ -333,14 +342,46 @@ private boolean checkTablePermission(ConnectorSecurityContext context, SchemaTab
         return false;
     }
 
+    private boolean checkTableAnyPermission(ConnectorSecurityContext context, SchemaTableName tableName)
+    {
+        if (INFORMATION_SCHEMA_NAME.equals(tableName.getSchemaName())) {
+            return true;
+        }
+        if (!isSchemaVisibleToUser(context, tableName.getSchemaName())) {
+            return false;
+        }
+        for (TableAccessControlRule rule : tableRules) {
+            Optional<Set<TablePrivilege>> tablePrivileges = rule.match(context.getIdentity().getUser(), tableName);
+            if (tablePrivileges.isPresent()) {
+                if (tablePrivileges.get().isEmpty()) {
+                    return false;
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
     private boolean isSchemaOwner(ConnectorSecurityContext context, String schemaName)
     {
         for (SchemaAccessControlRule rule : schemaRules) {
-            Optional<Boolean> owner = rule.match(context.getIdentity().getUser(), schemaName);
-            if (owner.isPresent()) {
-                return owner.get();
+            Optional<Set<SchemaPrivilege>> privileges = rule.match(context.getIdentity().getUser(), schemaName);
+            if (privileges.isPresent()) {
+                return privileges.get().contains(SchemaPrivilege.OWNERSHIP);
+            }
+        }
+        return false;
+    }
+
+    private boolean isSchemaVisibleToUser(ConnectorSecurityContext context, String schemaName)
+    {
+        for (SchemaAccessControlRule rule : schemaRules) {
+            Optional<Set<SchemaPrivilege>> privileges = rule.match(context.getIdentity().getUser(), schemaName);
+            if (privileges.isPresent()) {
+                return privileges.get().contains(SchemaPrivilege.OWNERSHIP) || privileges.get().contains(VISIBLE);
             }
         }
+        // if not rule match, we assume schema as invisible by default
```

Comment:
this is backward incompatible change, any existing installation without `SchemaAccessControlRule`s (with only `TableAccessControlRule`s) would break after the upgrade as user will have to modify access rules to add schema rules to existing table rules. I wonder we can avoid that.

Best to me would be to avoid adding `SchemaPrivilege` at all and rely only on existing rules. So I suggest to make schema visible if user has `owner=true` in `SchemaAccessControlRule` or exist any `TableAccessControlRule` that could give access to any table in given schema.

Reply:
> this is backward incompatible change, any existing installation without `SchemaAccessControlRule`s (with only `TableAccessControlRule`s) would break after the upgrade as user will have to modify access rules to add schema rules to existing table rules. I wonder we can avoid that.
> 
> Best to me would be to avoid adding `SchemaPrivilege` at all and rely only on existing rules. So I suggest to make schema visible if user has `owner=true` in `SchemaAccessControlRule` or exist any `TableAccessControlRule` that could give access to any table in given schema.

It's a valid concern. so we could assume schema is visible by default, if existing user use empty ` SchemaAccessControlRules ` 
aka on line 385 we return ` true ` if none ` SchemaRule ` match
@kokosing I will try to make it backward compatible